internal/kafka/internal/handlers/kafka_promote.go

```diff
@@ -0,0 +1,76 @@
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/internal/kafka/internal/services"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/errors"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/handlers"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/services/authorization"
+	"github.com/gorilla/mux"
+)
+
+type kafkaPromoteHandler struct {
+	service                        services.KafkaService
+	authService                    authorization.Authorization // TODO keep or remove this depending on the other TODOs of the kafka promote handlers
+	kafkaConfig                    *config.KafkaConfig
+	kafkaPromotionValidatorFactory KafkaPromotionValidatorFactory
+}
+
+func NewKafkaPromoteHandler(service services.KafkaService, authService authorization.Authorization, kafkaConfig *config.KafkaConfig, kafkaPromotionValidatorFactory KafkaPromotionValidatorFactory) *kafkaPromoteHandler {
+	return &kafkaPromoteHandler{
+		service:                        service,
+		authService:                    authService,
+		kafkaConfig:                    kafkaConfig,
+		kafkaPromotionValidatorFactory: kafkaPromotionValidatorFactory,
+	}
+}
+
+const kafkaPromoteRouteVariableID = "id"
+
+func (h kafkaPromoteHandler) Promote(w http.ResponseWriter, r *http.Request) {
+	var kafkaPromotionRequest public.KafkaPromotionRequest
+	id := mux.Vars(r)[kafkaPromoteRouteVariableID]
+	ctx := r.Context()
+	kafkaRequest, kafkaGetError := h.service.Get(ctx, id)
+	validateKafkaFound := func() handlers.Validate {
+		return func() *errors.ServiceError {
+			return kafkaGetError
+		}
+	}
+	cfg := &handlers.HandlerConfig{
+		MarshalInto: &kafkaPromotionRequest,
+		Validate: []handlers.Validate{
+			handlers.ValidateMinLength(&kafkaPromotionRequest.DesiredKafkaBillingModel, "desired_kafka_billing_model", 1),
+			// ValidateKafkaUserFacingUpdateFields(ctx, h.authService, kafkaRequest, &kafkaUpdateReq), // TODO should some of this checks in here related to owner be performed?
+			validateKafkaFound(),
+			validateKafkaRequestToPromoteHasAPromotableActualKafkaBillingModel(kafkaRequest),
+			validateKafkaRequestToPromoteHasAPromotableStatus(kafkaRequest),
+			validateRequestedKafkaPromotionHasDifferentKafkaBillingModel(&kafkaPromotionRequest, kafkaRequest),
+			validateNoKafkaPromotionInProgress(kafkaRequest),
+			validateKafkaPromotionRequestFields(&kafkaPromotionRequest, kafkaRequest, h.service, h.kafkaConfig, h.kafkaPromotionValidatorFactory),
+		},
+		Action: func() (i interface{}, serviceError *errors.ServiceError) {
+			kafkaFieldsToUpdate := map[string]interface{}{}
+			kafkaFieldsToUpdate["desired_kafka_billing_model"] = kafkaPromotionRequest.DesiredKafkaBillingModel
+			// We clear any potential promotion details being set due to a previous
+			// Kafka promotion failure
+			if kafkaRequest.PromotionDetails != "" {
+				kafkaFieldsToUpdate["promotion_details"] = ""
+			}
+			kafkaFieldsToUpdate["promotion_state"] = dbapi.KafkaPromotionStatePromoting
```

Comment:
If a kafka request is previously in `failed` state we allow performing a new promotion to allow API users to perform a promotion again